### PR TITLE
COMPAT: remove deprecated np.unicode_ and np.string_

### LINKED ISSUE #379: Compatibility with Numpy 2.0
I just noticed that pyogrio installed in geopandas' dev CI env, where we install nightly numpy 2 and pyogrio from source, does not work. We get this error

```py
A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.1.0.dev0 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "<string>", line 1, in <module>
  File "/home/runner/work/geopandas/geopandas/geopandas/tools/_show_versions.py", line 154, in show_versions
    proj_info = _get_C_info()
  File "/home/runner/work/geopandas/geopandas/geopandas/tools/_show_versions.py", line 61, in _get_C_info
    import pyogrio
  File "/home/runner/micromamba/envs/test/lib/python3.11/site-packages/pyogrio/__init__.py", line 9, in <module>
    from pyogrio.core import (
  File "/home/runner/micromamba/envs/test/lib/python3.11/site-packages/pyogrio/core.py", line 21, in <module>
    from pyogrio._io import ogr_list_layers, ogr_read_bounds, ogr_read_info
```

See the log within the Check and Log Environment section [here](https://github.com/geopandas/geopandas/actions/runs/8499094084/job/23279527137). I am not sure why is that as it is installed from source against numpy 2 but it seems to be something worth our attention.

### PR DESCRIPTION
**Pull request by martinfleis**
Closes #379 

GeoPandas CI, where we now test with pyogrio main and numpy nightly (after #381 allowed that), [caught](https://github.com/geopandas/geopandas/actions/runs/8550917712/job/23428851465) usage of deprecated `np.unicode_`. Replacing with `np.str_` as suggested by numpy.

~Shall I make an effort to add a CI env with nightly numpy?~ _edit: I did_

### CODE REVIEW

**Review by brendan-ward — APPROVED** [review]
Thanks @martinfleis !

One minor comment re: name of CI environment, but otherwise this looks good.

I'm not seeing any downsides to `np.str_` or `np.bytes_` listed in the numpy 2.0 migration guide, nor did our tests reveal any, so I'm assuming there are no gotchas there.

File: .github/workflows/tests-conda.yml [hunk 1]
```diff
@@ -40,6 +40,10 @@ jobs:
           - os: "windows-2019"
             python: "3.10"
             env: "libgdal3.5.1"
+          # environment with nightly wheels
+          - os: "ubuntu-latest"
+            python: "3.11"
+            env: "dev"
```
[the remark this answers is not in the document]
Comment (martinfleis) [reply .github/workflows/tests-conda.yml]: fair, renamed

File: pyogrio/tests/test_geopandas_io.py [hunk 2]
```diff
@@ -1446,6 +1446,7 @@ def test_write_nullable_dtypes(tmp_path):
     expected["col3"] = expected["col3"].astype("float32")
     expected["col4"] = expected["col4"].astype("float64")
     expected["col5"] = expected["col5"].astype(object)
+    expected.loc[1, "col5"] = None  # pandas 3.0.0+ converts to pd.NA on line above
```
Comment (jorisvandenbossche) [inline pyogrio/tests/test_geopandas_io.py:1449]:
```suggestion
    expected.loc[1, "col5"] = None  # pandas converts to pd.NA on line above
```

This is not something that should have changed in 3.0dev (I think current version of pandas do the same), but what did change is that the assert function became more strict in the exact missing value (None or pd.NA in an object dtype column are no longer considered as equal)

**Review by jorisvandenbossche — APPROVED** [review]
Thanks a lot, very good to start testing here with nightlies as well

**Comment by jorisvandenbossche** [comment]
Thanks!

**Pull request merged**